lib/internal/crypto/x509.js

```diff
@@ -0,0 +1,334 @@
+    }
+    return value;
+  }
+
+  get fingerprint() {
+    let value = this[kInternalState].get('fingerprint');
+    if (value === undefined) {
+      value = this[kHandle].fingerprint();
+      this[kInternalState].set('fingerprint', value);
+    }
+    return value;
+  }
+
+  get fingerprint256() {
+    let value = this[kInternalState].get('fingerprint256');
+    if (value === undefined) {
+      value = this[kHandle].fingerprint256();
+      this[kInternalState].set('fingerprint256', value);
+    }
+    return value;
+  }
+
+  get keyUsage() {
+    let value = this[kInternalState].get('keyUsage');
+    if (value === undefined) {
+      value = this[kHandle].keyUsage();
+      this[kInternalState].set('keyUsage', value);
+    }
+    return value;
+  }
+
+  get serialNumber() {
+    let value = this[kInternalState].get('serialNumber');
+    if (value === undefined) {
+      value = this[kHandle].serialNumber();
+      this[kInternalState].set('serialNumber', value);
+    }
+    return value;
+  }
+
+  get raw() {
+    let value = this[kInternalState].get('raw');
+    if (value === undefined) {
+      value = this[kHandle].raw();
+      this[kInternalState].set('raw', value);
+    }
+    return value;
+  }
+
+  get publicKey() {
+    let value = this[kInternalState].get('publicKey');
+    if (value === undefined) {
+      value = new PublicKeyObject(this[kHandle].publicKey());
+      this[kInternalState].set('publicKey', value);
+    }
+    return value;
+  }
+
+  toString() {
+    let value = this[kInternalState].get('pem');
+    if (value === undefined) {
+      value = this[kHandle].pem();
+      this[kInternalState].set('pem', value);
+    }
+    return value;
+  }
+
+  // There's no standardized JSON encoding for X509 certs so we
+  // fallback to providing the PEM encoding as a string.
+  toJSON() { return this.toString(); }
+
+  get ca() {
+    let value = this[kInternalState].get('ca');
+    if (value === undefined) {
+      value = this[kHandle].checkCA();
+      this[kInternalState].set('ca', value);
+    }
+    return value;
+  }
+
+  checkHost(name, options) {
+    validateString(name, 'name');
+    return this[kHandle].checkHost(name, getFlags(options));
+  }
+
+  checkEmail(email, options) {
+    validateString(email, 'email');
+    return this[kHandle].checkEmail(email, getFlags(options));
+  }
+
+  checkIP(ip, options) {
+    validateString(ip, 'ip');
+    return this[kHandle].checkIP(ip, getFlags(options));
+  }
+
+  checkIssued(otherCert) {
+    if (!isX509Certificate(otherCert))
+      throw new ERR_INVALID_ARG_TYPE('otherCert', 'X509Certificate', otherCert);
+    return this[kHandle].checkIssued(otherCert[kHandle]);
+  }
+
+  checkPrivateKey(pkey) {
+    if (!isKeyObject(pkey))
+      throw new ERR_INVALID_ARG_TYPE('pkey', 'KeyObject', pkey);
+    if (pkey.type !== 'private')
+      throw new ERR_INVALID_ARG_VALUE('pkey', pkey);
+    return this[kHandle].checkPrivateKey(pkey[kHandle]);
+  }
```

Comment:
I assume this compares the public key contained in the cert with the one derived from the provided private key?

There is a use-case in my mind coming from the JWK specification for checking that a certificate's public key is consistent with a public key obtained through different syntax. Would the underlying implementation accommodate if `pkey.type` was anything but `secret`?

Reply:
Yes, this checks that the given private key matches the certs public key.

For the public-to-public key check use case, we should likely add methods directly to `KeyObject` for checking that (in a separate PR)

Reply:
So isn't this method redundant with a future PR extending the KeyObject API in place? Given that one can get a public KeyObject using crypto.createPublicKey with privateKey as an argument

Reply:
Given that the other method doesn't exist yet, no ;-) ... if/when we get a method on `KeyObject` to check consistency we can update this to use that if necessary.